include/swift/IRGen/Linking.h

```diff
@@ -1553,6 +1556,16 @@ class LinkEntity {
   bool isAlwaysSharedLinkage() const;
 #undef LINKENTITY_GET_FIELD
 #undef LINKENTITY_SET_FIELD
+
+private:
+  static bool isObjCImplementation(NominalTypeDecl *NTD) {
```

Comment:
`hasObjCImplementation`, maybe?  But really I think this assertion can be broader — we should never have a type metadata or descriptor symbol for *any* imported class.  Unless we emit a lazy candidate with the normal symbol name, but I don't think we do?  And if we do, it would trigger for your case, too.